services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -86,59 +101,106 @@ public SubnetStateJson createSubnetState(@PathVariable String projectid, @Reques
+                } catch (Exception e) {
+                    throw new CompletionException(e);
+                }
+            }, ThreadPoolExecutorUtils.SELECT_POOL_EXECUTOR).handle((s, e) -> {
+                macResponseAtomic.set(s);
+                return s;
+            });
 
             // Verify VPC ID
-            VpcStateJson vpcResponse = this.subnetService.verifyVpcId(projectid, inSubnetState.getVpcId());
-            if (vpcResponse == null) {
-                throw new FallbackException("fallback request");
-            }
+            CompletableFuture<VpcStateJson> vpcFuture = CompletableFuture.supplyAsync(() -> {
+                try {
+                    return this.subnetService.verifyVpcId(projectid, inSubnetState.getVpcId());
+                } catch (Exception e) {
+                    throw new CompletionException(e);
+                }
+            }, ThreadPoolExecutorUtils.SELECT_POOL_EXECUTOR);
 
             //Prepare Route Rule(IPv4/6) for Subnet
-            RouteWebJson routeResponse = this.subnetService.createRouteRules(inSubnetState.getId(), inSubnetState);
-            if (routeResponse == null) {
-                throw new FallbackException("fallback request");
-            }
+            CompletableFuture<RouteWebJson> routeFuture = CompletableFuture.supplyAsync(() -> {
+                try {
+                    return this.subnetService.createRouteRules(inSubnetState.getId(), inSubnetState);
+                } catch (Exception e) {
+                    throw new CompletionException(e);
+                }
+            }, ThreadPoolExecutorUtils.SELECT_POOL_EXECUTOR).handle((s, e) -> {
+                routeResponseAtomic.set(s);
+                return s;
+            });;
+
 
-            //Allocate Gateway Mac
-//            MacStateJson macResponse = this.subnetService.allocateMacGateway(projectid, inSubnetState.getVpcId(), portId);
-//            if (macResponse == null) {
-//                throw new ResourcePersistenceException();
-//            }
 
             // Verify/Allocate Gateway IP, subnet id, port id, subnet cidr, response:IP - unique
 //            IPStateJson ipResponse = this.subnetService.allocateIPGateway(inSubnetState.getId(), inSubnetState.getCidr(), portId);
 //            if (ipResponse == null) {
 //                throw new ResourcePersistenceException();
 //            }
 
+            // Synchronous blocking
+            CompletableFuture<Void> allFuture = CompletableFuture.allOf(vpcFuture, macFuture, routeFuture);
+            allFuture.join();
+
+            macResponse = macFuture.join();
+            routeResponse = routeFuture.join();
+            logger.info("Total processing time:" + (System.currentTimeMillis() - start) + "ms");
+
             // set up value of properties for subnetState
-            List<RouteWebObject> routes = subnetState.getRoutes();
+            List<RouteWebObject> routes = inSubnetState.getRoutes();
```

Comment:
what would happen if inSubnetState includes some route rules, and we add some more from routeResponse which could be duplicated? Will that create multiple copies of the same rules? like so as we are using List

Reply:
Can we skip inSubnetState.getRoutes()

Reply:
Yeah, if inSubnetState includes some route rules already, we will just add the rules just created into the list. Do you have any good suggestion?